android: add critical section to prevent concurrency

jasaw opened the pull request:
Protect connect and disconnect method calls and gatt callback from concurrency by wrapping the code in critical section.

jasaw commented:
@chipweinberger Concurrency is an existing problem in FBP. I think critical section is still the cleanest way of addressing it.

chipweinberger commented:
Yes they are concurrent

but what is the actual bug it would cause?

chipweinberger commented:
also, i think the code will look nicer if

1. we grab the local every time we run the method call handler

2. call it mMethodCallMutex


I'm not a big fan of the CriticalSection abstraction. it's a good idea though 

I think by splitting the PRs we will have a diff that I can be confident in

jasaw commented:
>what is the actual bug it would cause?

Just analyze this chunk of code in gatt callback for illustration purpose:

```
                if(newState == BluetoothProfile.STATE_DISCONNECTED) {

                    // remove from connected devices
                    mConnectedDevices.remove(remoteId);

                    // remove from currently connecting devices
                    mCurrentlyConnectingDevices.remove(remoteId);

                    // remove from currently bonding devices
                    mBondingDevices.remove(remoteId);

                    // we cannot call 'close' for autoconnected devices
                    // because it prevents autoconnect from working
                    if (mAutoConnected.containsKey(remoteId)) {
                        log(LogLevel.DEBUG, "autoconnect is true. skipping gatt.close()");
                    } else {
                        // it is important to close after disconnection, otherwise we will 
                        // quickly run out of bluetooth resources, preventing new connections
                        gatt.close();
                    }
                }
```

If connect method call executes in between `mBondingDevices.remove(remoteId)` and `gatt.close()`, we end up with a situation where `gatt.close()` is called right after `connectGatt` call, which will result in stranded Bluetooth connection race condition that we worked around with a 2s delay in dart, except this time, it won't even go to dart layer, so our 2s delay won't help us.


>just curious, as we've never had a bug report caused by this yet

Do other people pay close enough attention to FBP logs to notice the absence of gatt callback while ending up with an established Bluetooth connection? I'm not sure. I've been using FBP and original FB for years and didn't notice this until recently when I push the connect/disconnect a bit harder.

chipweinberger commented:
thanks for analyzing it.

yup, that's a pretty clear potential issue!

jasaw commented:
@chipweinberger The diff looks a lot nicer now. Thoughts?

chipweinberger commented:
certainly nicer.

but i think i prefer

1. don't use CriticalSection abstraction

2. use the existing try/catch/finally in method call handler for acquire/release

3. maybe let's ignore the chance of the mutex not being acquired (this should be researched more) / or create a new simple helper function for acquire with retry

jasaw commented:
If we replace CriticalSection with try/catch/finally & acquire/release, we're going to end up with duplicated boilerplate code on both method call handler and gatt callback. Is this really the cleanest approach? Just want a confirmation before I proceed with code changes.

Another thing to consider for the future is to consider wrapping other callback function in critical section.

chipweinberger commented:
i think it would look clean 

you can use this i think

semaphore.tryAcquire(Long.MAX_VALUE, TimeUnit.DAYS));

it won't throw

chipweinberger commented:
and the interrupted exception only happens if 

thread.interrupt() is called afaict 

so it's unclear if it would happen

jasaw commented:
>maybe we should just make our own helper function to acquire a mutex with retry

It's definitely easier to implement our own helper function that retries and this is what I've implemented now.

I've also replaced critical section with try/catch/finally.

chipweinberger commented:
looking much better now

chipweinberger commented on android/src/main/java/com/lib/flutter_blue_plus/FlutterBluePlusPlugin.java:
```diff
@@ -2091,62 +2110,76 @@ public void onScanFailed(int errorCode)
-                return;
-            }
+            try {
+                acquireMethodCallMutex();
```
add comment here

jasaw replied:
Done.

jasaw commented on android/src/main/java/com/lib/flutter_blue_plus/FlutterBluePlusPlugin.java:
```diff
@@ -2091,62 +2110,76 @@ public void onScanFailed(int errorCode)
+
+                invokeMethodUIThread("OnConnectionStateChanged", response);
+
+            } catch (Exception e) {
```
I'm not sure what the behavior is if we just let it crash. Which is a nicer solution? let it crash or log it?

chipweinberger replied:
crash

jasaw replied:
Done.

jasaw commented on android/src/main/java/com/lib/flutter_blue_plus/FlutterBluePlusPlugin.java:
```diff
@@ -170,6 +172,19 @@ else if (ends)
         }    
     }
 
+    private void acquireMethodCallMutex()
```
Done.

chipweinberger commented:
seems ok now

please squash and test 

thanks

jasaw commented:
Squash and tested.

chipweinberger commented:
merged!
